### PR TITLE
feat: instrumentation for chromadb

### PR DESCRIPTION
**Pull request by 5war00p**
/claim #78

### CODE REVIEW

File: packages/traceloop-sdk/src/lib/tracing/index.ts [hunk 1]
```diff
@@ -301,6 +329,60 @@ export const startTracing = (options: InitializeOptions) => {
   });
 
   _sdk.start();
+
```
Comment (nirga) [inline packages/traceloop-sdk/src/lib/tracing/index.ts]: I think it's a leftover from when you merged to `main`?

File: packages/sample-app/src/sample_chromadb.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
+import * as traceloop from "@traceloop/node-server-sdk";
```
Comment (nirga) [inline packages/sample-app/src/sample_chromadb.ts:1]: We try to make these examples resemble "real-life" scenarios, can you do that here as well? (Similar to the sample app we have in Python)

File: packages/instrumentation-chromadb/tests/instrumentation.test.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+import { Polly, setupMocha as setupPolly } from "@pollyjs/core";
+import FetchAdapter from "@pollyjs/adapter-fetch";
+import FSPersister from "@pollyjs/persister-fs";
+
+const memoryExporter = new InMemorySpanExporter();
+
+Polly.register(FetchAdapter);
+Polly.register(FSPersister);
+
+describe.skip("Test Pinecone instrumentation", function () {
```
Comment (nirga) [inline packages/instrumentation-chromadb/tests/instrumentation.test.ts]:
```suggestion
describe.skip("Test Chroma instrumentation", function () {
```

File: packages/instrumentation-chromadb/tests/instrumentation.test.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+  setupPolly({
+    adapters: ["fetch"],
+    persister: "fs",
+    recordIfMissing: process.env.RECORD_MODE === "NEW",
+  });
+
+  before(async () => {
+    if (process.env.RECORD_MODE !== "NEW") {
+      process.env.PINECONE_API_KEY = "test";
```
Comment (nirga) [inline packages/instrumentation-chromadb/tests/instrumentation.test.ts]: remove

File: packages/instrumentation-chromadb/tests/instrumentation.test.ts
Comment (nirga) [inline packages/instrumentation-chromadb/tests/instrumentation.test.ts:1]: Can you add tests here?

File: packages/instrumentation-chromadb/src/instrumentation.ts [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    const attributes: Attributes = {
+      [SpanAttributes.VECTOR_DB_VENDOR]: "ChromaDB",
+    };
+    const span = this.tracer.startSpan(`chromadb.${methodName}`, {
+      kind: SpanKind.CLIENT,
+      attributes,
+    });
+
+    // Instrumenting only for query and peak
+    if (methodName === "query" || methodName === "pek") {
```
Comment (nirga) [inline packages/instrumentation-chromadb/src/instrumentation.ts]: seems like a typo here? `pek`?

File: packages/traceloop-sdk/src/lib/tracing/index.ts (hunk 1 above)
Comment (5war00p) [reply packages/traceloop-sdk/src/lib/tracing/index.ts]: Shall I remove it?

File: packages/instrumentation-chromadb/src/instrumentation.ts (hunk 5 above)
Comment (5war00p) [reply packages/instrumentation-chromadb/src/instrumentation.ts]: Oops wil fix it

File: packages/sample-app/src/sample_chromadb.ts (hunk 2 above)
Comment (5war00p) [reply packages/sample-app/src/sample_chromadb.ts:1]: Sure

File: packages/traceloop-sdk/src/lib/tracing/index.ts (hunk 1 above)
Comment (nirga) [reply packages/traceloop-sdk/src/lib/tracing/index.ts]: Ofc, unless you meant to add it?

Comment (5war00p) [reply packages/traceloop-sdk/src/lib/tracing/index.ts]: I see it still exists in the main branch.

Comment (5war00p) [reply packages/traceloop-sdk/src/lib/tracing/index.ts]: Looks like not properly rebased this file, nvm, i'm doing it again

**Comment by 5war00p** [comment]
image
image

File: .gitignore [hunk 6]
```diff
@@ -44,4 +44,6 @@ Thumbs.db
 .nx
 
 # env
-.env*
\ No newline at end of file
+.env*
+/packages/sample-app/chromadb
```
Comment (nirga) [inline .gitignore]: why?

File: packages/instrumentation-chromadb/package.json [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    "@opentelemetry/semantic-conventions": "^1.22.0",
+    "@traceloop/ai-semantic-conventions": "^0.6.0"
+  },
+  "devDependencies": {
+    "@pollyjs/adapter-node-http": "^6.0.6",
+    "@pollyjs/core": "^6.0.6",
+    "@pollyjs/persister-fs": "^6.0.6",
+    "chromadb": "^1.8.1"
+  },
+  "homepage": "https://github.com/traceloop/openllmetry-js/tree/main/packages/instrumentation-openai",
```
Comment (nirga) [inline packages/instrumentation-chromadb/package.json]:
```suggestion
  "homepage": "https://github.com/traceloop/openllmetry-js/tree/main/packages/instrumentation-chromadb",
```

File: packages/instrumentation-chromadb/README.md
Comment (nirga) [inline packages/instrumentation-chromadb/README.md:1]: Make sure this README is for ChromaDB (right now it contains VertexAI, Cohere, and OpenAI)

File: packages/instrumentation-chromadb/src/instrumentation.ts [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+      ["^1.8.1"],
+      this.wrap.bind(this),
+      this.unwrap.bind(this),
+    );
+
+    return module;
+  }
+
+  public manuallyInstrument(module: typeof chromadb) {
+    return module;
```
Comment (nirga) [inline packages/instrumentation-chromadb/src/instrumentation.ts]: You should call `wrap` here

File: packages/instrumentation-chromadb/src/instrumentation.ts [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+      [SpanAttributes.VECTOR_DB_VENDOR]: "ChromaDB",
+    };
+    const span = this.tracer.startSpan(`chromadb.${methodName}`, {
+      kind: SpanKind.CLIENT,
+      attributes,
+    });
+
+    // Instrumenting only for query and peak
+    if (methodName === "query" || methodName === "peek") {
+      const query_request_event = span.addEvent("chromadb.query.request");
```
Comment (nirga) [inline packages/instrumentation-chromadb/src/instrumentation.ts]: Do not add vectors if `traceContent` config is `false`

File: packages/instrumentation-chromadb/src/instrumentation.ts [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+    const attributes: Attributes = {
+      [SpanAttributes.VECTOR_DB_VENDOR]: "ChromaDB",
+    };
+    const span = this.tracer.startSpan(`chromadb.${methodName}`, {
+      kind: SpanKind.CLIENT,
+      attributes,
+    });
+
+    // Instrumenting only for query and peak
+    if (methodName === "query" || methodName === "peek") {
```
Comment (nirga) [inline packages/instrumentation-chromadb/src/instrumentation.ts]: Events should match [the format we have in Python](https://github.com/traceloop/openllmetry/blob/main/packages/opentelemetry-instrumentation-pinecone/opentelemetry/instrumentation/pinecone/__init__.py) - each vector in a separate event, etc.

File: .gitignore (hunk 6 above)
Comment (5war00p) [reply .gitignore]: So when we run chromadb locally it creates the files do we need to commit those? It should be local right?

File: packages/instrumentation-chromadb/src/instrumentation.ts (hunk 10 above)
Comment (5war00p) [reply packages/instrumentation-chromadb/src/instrumentation.ts]: @nirga I've added event based attributes similar to Python. Please re-review it.

image

File: .gitignore (hunk 6 above)
Comment (nirga) [inline .gitignore]: This is not a good thing to put in a gitignore. It's better to put the chromadb in some known folder with a distinct name like ".temp-chroma"

File: packages/instrumentation-chromadb/tests/instrumentation.test.ts [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+    assert.strictEqual(
+      attributes["db.chroma.upsert.documents_count"],
+      JSON.stringify(updatedInput?.documents?.length),
+    );
+
+    // Cleanup created collection at the end
+    collection.delete({ ids: ["uri7", "uri8"] });
+  });
+
+  it.skip("should set span attributes for Modify", async () => {
```
Comment (nirga) [inline packages/instrumentation-chromadb/tests/instrumentation.test.ts]: why is it in `skip`?

File: packages/instrumentation-chromadb/tests/instrumentation.test.ts [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+
+describe("Test ChromaDB instrumentation", function () {
+  const provider = new BasicTracerProvider();
+  let instrumentation: ChromaDBInstrumentation;
+  let contextManager: AsyncHooksContextManager;
+  let chromaDbClient: chromadb.ChromaClient;
+  let collection: chromadb.Collection;
+  let chromaRun: ChildProcess;
+
+  setupPolly({
```
Comment (nirga) [inline packages/instrumentation-chromadb/tests/instrumentation.test.ts]: Why do you need to setup Polly?

Comment (5war00p) [reply packages/instrumentation-chromadb/tests/instrumentation.test.ts]: removed it

File: packages/instrumentation-chromadb/tests/instrumentation.test.ts (hunk 11 above)
Comment (5war00p) [reply packages/instrumentation-chromadb/tests/instrumentation.test.ts]: unskipped these

File: packages/instrumentation-chromadb/tests/instrumentation.test.ts [hunk 13]
```diff
@@ -0,0 +1,380 @@
...
+  let instrumentation: ChromaDBInstrumentation;
+  let contextManager: AsyncHooksContextManager;
+  let chromaDbClient: chromadb.ChromaClient;
+  let collection: chromadb.Collection;
+  let chromaRun: ChildProcess;
+
+  this.beforeAll(async () => {
+    // Run ChromaDB instance on different terminal instance
+    chromaRun = spawn("/bin/sh");
+    chromaRun.stdin?.write("chroma run --path .\n");
```
Comment (nirga) [inline packages/instrumentation-chromadb/tests/instrumentation.test.ts:44]: can we use `exec` here instead?

File: packages/instrumentation-chromadb/tests/instrumentation.test.ts [hunk 14]
```diff
@@ -0,0 +1,380 @@
...
+      attributes["db.chroma.query.where_document"],
+      JSON.stringify(input?.whereDocument),
+    );
+    assert.strictEqual(
+      attributes["db.chroma.query.include"],
+      JSON.stringify(input?.include),
+    );
+
+    const events = spans[1].events;
+    assert.strictEqual(events[0].name, "db.chroma.request");
```
Comment (nirga) [inline packages/instrumentation-chromadb/tests/instrumentation.test.ts]: can you assert on the values of the events?

File: packages/instrumentation-chromadb/src/instrumentation.ts [hunk 15]
```diff
@@ -0,0 +1,345 @@
...
+      ["^1.8.1"],
+      this.wrap.bind(this),
+      this.unwrap.bind(this),
+    );
+
+    return module;
+  }
+
+  public manuallyInstrument(module: typeof chromadb) {
+    this._diag.debug(`Manually patching chromadb-ai`);
```
Comment (nirga) [inline packages/instrumentation-chromadb/src/instrumentation.ts]:
```suggestion
    this._diag.debug(`Manually patching chromadb`);
```

File: .gitignore (hunk 6 above)
Comment (5war00p) [reply .gitignore]: We no longer required this. And we are gonna directly ignore `chroma.sqllite3`

File: packages/instrumentation-chromadb/tests/instrumentation.test.ts (hunk 14 above)
Comment (5war00p) [reply packages/instrumentation-chromadb/tests/instrumentation.test.ts]: Done

File: packages/instrumentation-chromadb/tests/instrumentation.test.ts (hunk 13 above)
Comment (5war00p) [reply packages/instrumentation-chromadb/tests/instrumentation.test.ts:44]: Done

**Review by nirga — APPROVED** [review]

**Pull request merged**